nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/P2FacetImpl.java

```diff
@@ -0,0 +1,283 @@
+      tx.saveAsset(asset);
+    }
+
+    return asset;
+  }
+
+  /**
+   * Find a component by its name and tag (version)
+   *
+   * @return found component of null if not found
+   */
+  @Nullable
+  public Component findComponent(final StorageTx tx,
+                                        final Repository repository,
+                                        final String name,
+                                        final String version)
+  {
+    Iterable<Component> components = tx.findComponents(
+        Query.builder()
+            .where(P_NAME).eq(name)
+            .and(P_VERSION).eq(version)
+            .build(),
+        singletonList(repository)
+    );
+    if (components.iterator().hasNext()) {
+      return components.iterator().next();
+    }
+    return null;
+  }
+
+  /**
+   * Find an asset by its name.
+   *
+   * @return found asset or null if not found
+   */
+  @Nullable
+  public Asset findAsset(final StorageTx tx, final Bucket bucket, final String assetName) {
+    return tx.findAssetWithProperty(MetadataNodeEntityAdapter.P_NAME, assetName, bucket);
+  }
+
+  /**
+   * Save an asset && create blob.
+   *
+   * @return blob content
+   */
+  public Content saveAsset(final StorageTx tx,
+                                  final Asset asset,
+                                  final Supplier<InputStream> contentSupplier,
+                                  final Payload payload) throws IOException
+  {
+    AttributesMap contentAttributes = null;
+    String contentType = null;
+    if (payload instanceof Content) {
+      contentAttributes = ((Content) payload).getAttributes();
+      contentType = payload.getContentType();
+    }
+    return saveAsset(tx, asset, contentSupplier, contentType, contentAttributes);
+  }
+
+  /**
+   * Save an asset && create blob.
+   *
+   * @return blob content
+   */
+  public Content saveAsset(final StorageTx tx,
+                                  final Asset asset,
+                                  final Supplier<InputStream> contentSupplier,
+                                  final String contentType,
+                                  @Nullable final AttributesMap contentAttributes) throws IOException
+  {
+    Content.applyToAsset(asset, Content.maintainLastModified(asset, contentAttributes));
+    AssetBlob assetBlob = tx.setBlob(
+        asset, asset.name(), contentSupplier, HASH_ALGORITHMS, null, contentType, false
+    );
+    asset.markAsDownloaded();
+    tx.saveAsset(asset);
+    return toContent(asset, assetBlob.getBlob());
+  }
+
+  /**
+   * Convert an asset blob to {@link Content}.
+   *
+   * @return content of asset blob
+   */
+  public Content toContent(final Asset asset, final Blob blob) {
+    Content content = new Content(new BlobPayload(blob, asset.requireContentType()));
+    Content.extractFromAsset(asset, HASH_ALGORITHMS, content.getAttributes());
+    return content;
+  }
+
+  @Override
+  public AssetKind getAssetKind(final String path) {
```

Comment:
add unit tests tor this function.
@proteus1121  please recheck the logic of asset kind discovery

Reply:
looks same as in P2ProxyRecipe.configure()